### PR TITLE
Automate the generation changelogs in text format

### PR DESCRIPTION
**Pull request by soham30rane**
Fixes [#471](https://github.com/sagemath/website/issues/471) as discussed on [sage-devel](https://groups.google.com/g/sage-devel/c/1dRyPFPGEKc) 2024-11-13

Necessary steps to take before achieving proper functionality:
 -  Create Personal Access Token and add it to **repo > Settings > Secrets and Variables > Actions > Repository Secrets**
with name `PERSONAL_ACCESS_TOKEN`
 -  Add another secret with name `CHANGELOG_TRIGGER_SECRET`.
 -  Also add both the secrets to [sage repo](https://github.com/sagemath/sage) with the exact same names and values.
 
 I am also making a pr to [sage repo](https://github.com/sagemath/sage) to define a workflow which will in turn trigger `generate_changelog.yml` on every release.
 
 Sample results of the script. (Executed it for release tag 10.5, which considered all pre-releases uptil now)
 [sage-10.5.txt](https://github.com/user-attachments/files/17879837/sage-10.5.txt)

### CODE REVIEW

File: scripts/create_changelog.py
Comment (kwankyu) [inline scripts/create_changelog.py:186]: It is a slight deviation from the tradition, but it may be better to use `(http://www.sagemath.org)` with no trailing slash for a new start.

File: scripts/create_changelog.py
Comment (kwankyu) [inline scripts/create_changelog.py:191]: `[First contribution]` to align with other bracketed phrases.

**Comment by kwankyu** [comment]
Can I see an example run of the workflow?

**Comment by dimpase** [comment]
I wonder about the need for PAT here. Can it be explained?

**Comment by soham30rane** [comment]
> Can I see an example run of the workflow?

[This commit](https://github.com/sagemath/website/pull/480/commits/38ccbd32ec77894d8ae7528c8c8794539d2d947f) is an example run of the workflow.
I triggered it mannually with the release tag of 10.0.
You could have a look [here](https://github.com/soham30rane/sagemath-website/actions). The previous trigger failed because I forgot to add `PERSONAL_ACCESS_TOKEN` to the secret.

Actually the workflow in this repo is expected to be triggered by [this workflow](https://github.com/sagemath/sage/pull/39024/commits/0fe4a58321fae3d114e234b81e771af0a5d1fe79) located in main sage repo

**Comment by soham30rane** [comment]
> I wonder about the need for PAT here. Can it be explained?

PAT is required for using Github REST API

**Need for Github REST API**: Everytime a release occurs on main sage repo, it triggers the generation of changelog in this repo. To generate changelog, the script uses info from the release notes (which is fetched via Github API). 
But even if we somehow manage to get the release note info available to the script without using the API, it is not enough, because it doesn't contain the reviewer info and multiple authors info for PRs. 
Hence Github API is needed to individually fetch info about each of the PRs.

File: scripts/create_changelog.py
Comment (soham30rane) [reply scripts/create_changelog.py:186]: Done!

File: scripts/create_changelog.py
Comment (soham30rane) [reply scripts/create_changelog.py:192]: Done [here](https://github.com/sagemath/website/pull/480/commits/f96158d146f4e554b88e3f074e278ff8602e91aa)!

**Comment by dimpase** [comment]
But isn't PAT only needed for the user running this particular Actions workflow?
This user need not be the user who submitted their PR. In our workflow, it can be a maintainer.

This action can be run by someone who has rights to approve the PR changing contributor's info. Other users don't seem to need a PAT, right?

**Comment by soham30rane** [comment]
> Other users don't seem to need a PAT, right?

Yes, you are right, only a single PAT needs to be created (by a maintainer) and stored in repository secrets.

> This action can be run by someone who has rights to approve the PR changing contributor's info

Yes! And I just realized that we don't need `CHANGELOG_TRIGGER_SECRET` to verify the source of trigger, because the PAT is already ensuring that only those people who have rights to approve the PR can trigger the workflow via API call. Should I go ahead and do the related cleanup?

**Comment by dimpase** [comment]
I admit don't have a complete understanding of Actions permissions model - please proceed as you see fit.

File: .github/workflows/generate_changelog.yml
Comment (kwankyu) [inline .github/workflows/generate_changelog.yml:50]: Could the user.name and user.email be simpler?

We may use a specific file name (`src/changelogs/sage-......txt`?) instead of `.` for safety?

File: scripts/create_changelog.py
Comment (kwankyu) [inline scripts/create_changelog.py:9]: `add it to`

File: scripts/create_changelog.py
Comment (kwankyu) [inline scripts/create_changelog.py:160]: If we get recent tags in the first page,  I think 100 would be enough.

File: src/changelogs/sage-10.0.txt
Comment (kwankyu) [inline src/changelogs/sage-10.0.txt:377]: Previous changelogs do not have an empty line at the end.

File: scripts/create_changelog.py
Comment (kwankyu) [inline scripts/create_changelog.py:67]: It seems that we can retrieve the github user's real name via github api from his/her github account name. If this is the case, then we may create a new entry in `conf/contributors.xml` and then use the real name instead of the github account name in the created changelog. For example, we may create this entry
```
<contributor
 name="Marie Bonboire"
 github="marizee"/>
```
and use "Marie Bonboire" instead of `@marizee`. The entries in `contributors.xml` are listed in alphabetical order in the last name. We may use the fact to find the right place to put the new entry in.

But this is out of the scope of this PR. You may decide to do this extra work or not. Just let me know.

**Comment by kwankyu** [comment]
> ... only a single PAT needs to be created (by a maintainer) and stored in repository secrets.

It seems that sagemath organization owner (not the admin of sagemath/sage) can create a PAT that allows (read)-access to sagemath/sage repo, and then the admin of sagemath/website should store the PAT as a secret of sagemath/sage repo.

Do you confirm this?

**Comment by soham30rane** [comment]
Yes, I confirm that the sagemath organization owner can create a PAT, that allows read-access to sagemath/sage (since sagemath/sage is a public repo, giving 'read access to public repos' is also enough). BUT the PAT needs to be stored as a secret for both repos sagemath/website and sagemath/sage.

I will try to elaborate on the use of PAT here: (Its used in two parts)

1. In  `create_changelog.py` which will be located in sagemath/website. Here its used to retrieve PR related info using Github API

2. In [this workflow](https://github.com/sagemath/sage/pull/39024/commits/0fe4a58321fae3d114e234b81e771af0a5d1fe79) which will be located in sagemath/sage. Here its used to trigger  `generate_changelog.yml` workflow after each release via Github API which will generate the changelog.

So for the 1st use case any valid PAT with 'read-access to public repos' will work. (And it should be stored in sagemath/website)
For 2nd use case, the PAT used has to belong to person who has rights to trigger a workflow run in sagemath/website (Obviously the org owner will have those rights). (And it should be stored in sagemath/sage)

This is the reason we need to store PAT in secrets of both repos sagemath/website and sagemath/sage. So we can store the PAT created by org owner (that allows read access to public repos) in secrets of both repos.

File: .github/workflows/generate_changelog.yml
Comment (soham30rane) [reply .github/workflows/generate_changelog.yml:50]: > Could the user.name and user.email be simpler?

Yes we could do that, but if we change the email then we won't see the black github-actions icon github-actions in the commits. Instead we would see it in gray disabled mode.

>   We may use a specific file name (src/changelogs/sage-......txt?) instead of . for safety?

Yes, will fix that

File: scripts/create_changelog.py
Comment (soham30rane) [reply scripts/create_changelog.py:67]: Yes, this is doable, but some users might not have their real name on their Github Profile, in such case we seem to have no other option than to user their github username in the format `@github-username`.
I will work on it

**Comment by kwankyu** [comment]
For public repos, no PAT may be necessary (or so my AI says :-) We could experiment soon.

**Comment by soham30rane** [comment]
I tried running the workflow without a PAT, and it worked perfectly! Sorry for the confusion earlier, also thank you for your patience.
That said, for the second use case I mentioned, a PAT is still necessary. If you'd prefer to avoid creating a PAT, you could just sideline it and manually trigger `generate_changelog.yml` after every release. Whatever works best for the org!

**Comment by soham30rane** [comment]
Okay, so the last time I triggered the workflow on my testing repo I had put a limit on the number of PRs fetched (just for the sake of saving time) and it seemed to work perfectly fine.
Today I ran the workflow after removing that limit and it gave me 'rate limit reached'.
I skimmed through their [official documentation](https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28), and here they say `The primary rate limit for unauthenticated requests is 60 requests per hour`. This is not enough as there are more than 60 PRs in any stable release

**Comment by kwankyu** [comment]
So with a PAT, the rate limit is 5000 PRs? This would be big enough.

**Comment by soham30rane** [comment]
Yeah, 5000 would be more than enough.

**Comment by haraldschilly** [comment]
I'm not sure what's going on and how this is supposed to work, but I just opened the settings and saw there are also "Organization secrets" (and there is one for PYPI).

When I open the management form for that secret, I can add one there and tell it exactly which repositories should have access. So, I'm only guessing, but maybe you want the same secret (with the same value) to appear in two repositories? Then we could make this such an org secret for two repositories.

**Comment by kwankyu** [comment]
We can proceed in this way: Harald, as you are the organization owner, you can designate me as an admin of sagemath/website. Then I can work with the author to put his code in the repo, and then we can test how it works. And then if we need a PAT (perhaps but I am not sure yet), then you, as an organization owner, could make one for us. Then I can put that PAT as a secret to either repo sagemath/website or sagemath/sage as necessary.

**Comment by dimpase** [comment]
I think it's better to have different PATs for different repos, just as with the usual passwords

**Comment by soham30rane** [comment]
Major changes in recent commits:
 - Removed unnecessary secret `CHANGELOG_TRIGGER_SECRET`. It was supposed to verify the source of Trigger. But the PAT already verifies it.
 - If any contributor is not found in `conf/contributors.xml`, then their name is fetched from their github profile (if available). Then their entry is added in `conf/contributors.xml`.
 
 I tried to ensure that the entries in `conf/contributors.xml` are sorted by alphabetical order in the last name, but faced a minor issue.
  - My approach assumed that the last name of a contributor is the last word in their name. This assumption worked in almost all cases
  - Very few contributors have more than 2 words in their name and their last name is not actually the last word (e.g. Luis Felipe Tabera Alonso). This caused the ordering to slightly deviate from the original arrangement in `conf/contributors.xml`

**Comment by kwankyu** [comment]
> * Very few contributors have more than 2 words in their name and their last name is not actually the last word (e.g. Luis Felipe Tabera Alonso). This caused the ordering to slightly deviate from the original arrangement in `conf/contributors.xml`

OK. Good.

**Comment by dimpase** [comment]
> [This commit](https://github.com/sagemath/website/pull/480/commits/38ccbd32ec77894d8ae7528c8c8794539d2d947f) is result of an example [run of the workflow](https://github.com/soham30rane/sagemath-website/actions/runs/11997349934).

I looked at the resulting list there and noticed that some GitHub usernames are not replaced by real names, e.g. `@fingolfin` is not given as `Max Horn` (the name you can see on https://github.com/fingolfin/)

File: conf/contributors.xml
Comment (kwankyu) [inline conf/contributors.xml:1360]: Same person with the previous one? May combine?

You may use ```altnames="Ingólfur Eðvarðsson"```.

File: scripts/create_changelog.py
Comment (kwankyu) [reply scripts/create_changelog.py:283]: Not yet.

File: .github/workflows/generate_changelog.yml
Comment (kwankyu) [inline .github/workflows/generate_changelog.yml:27]: I think this name `PERSONAL_ACCESS_TOKEN` is too general. How about `SAGE_ACCESS_TOKEN`?

File: scripts/create_changelog.py
Comment (kwankyu) [inline scripts/create_changelog.py:52]: Would you add short docstrings to functions (except simple ones), in particular explaining the arguments? They may be obvious now, but later we may forget the intention.

File: conf/contributors.xml
Comment (dimpase) [reply conf/contributors.xml:1360]: it's the same person, from Iceland. My former student

File: scripts/create_changelog.py
Comment (kwankyu) [inline scripts/create_changelog.py:74]: This sorts the list of contributors by "last name" all again each time when a release is made. This may destroy the correct order of some contributors made manually since their last names are special.

How about inserting each of new contributors into the tree finding the correct place while keeping the order of existing contributors? This would be a bit more efficient as well (though efficiency is a least concern here...)

You may use the same sort key `key=lambda x: unidecode(get_last_name(x.get('name', '\x7F'))` to find the right place for each contributor.

File: scripts/create_changelog.py
Comment (kwankyu) [inline scripts/create_changelog.py:125]: this print is for debugging?

File: scripts/create_changelog.py
Comment (kwankyu) [inline scripts/create_changelog.py:175]: If `release_data` is `None`, this function should not be called in the first place. Perhaps this `if` statement should be moved to the caller?

File: scripts/create_changelog.py
Comment (kwankyu) [inline scripts/create_changelog.py:31]: `contributor name`

File: scripts/create_changelog.py
Comment (kwankyu) [inline scripts/create_changelog.py:38]: `of first constributors`?

File: scripts/create_changelog.py
Comment (kwankyu) [reply scripts/create_changelog.py:283]: Or perhaps reverted by mistake?

**Comment by kwankyu** [comment]
I think I have sorted out how the PAT things work, with fingers crossed.

We need two PATs:

1. SAGE_ACCESS_TOKEN: created by a person with read-permission to sagemath/sage, stored in sagemath/website as a secret. This PAT is used by the workflow `Generate Changelog` to access release data from sagemath/sage.

2. WEBSITE_ACCESS_TOKEN: created by a person with write-permission to sagemath/website, stored in sagemath/sage as a secret. This PAT is used by the workflow `Trigger Changelog Generation` (https://github.com/sagemath/sage/pull/39024) to trigger the workflow `Generate Changelog` at a release time.

It is not advisable to use PATs created by someone with more-than-necessary permissions. Hence I am excluded.

We may proceed in the following schedule, after this PR gets merged to this repo:

1. Dima creates SAGE_ACCESS_TOKEN (classic PAT, No expiration, "repo" scope) and I store it as a secret to sagemath/website.
2. I test the workflow `Generate Changelog` by manually running it for versions 10.1,10.2,10.3,10.4
3. I review PR #39024 and wait for it to be merged.
4. I give Dima write-permission to sagemath/website.
5. Dima creates WEBSITE_ACCESS_TOKEN(classic PAT, No expiration, "repo" and "workflow" scopes) and I store it as a secret to sagemath/sage.
6. I test the workflow `Trigger Changelog Generation` by manually running it for version 10.5

**Comment by dimpase** [comment]
OK - but not now - it's 22:50 local (Chicago) time for me. (We have moved, to Northwestern U.)

**Comment by kwankyu** [comment]
Congratulations!

File: scripts/create_changelog.py
Comment (soham30rane) [reply scripts/create_changelog.py:74]: I did consider this approach. But this approach seemed to have a significant limitation.
For example, consider the following existing order in `contributors.xml`
```
<contributor name="John Doe"/>
<contributor name="Jane Ann Doe Zander"/>
..
..
```
Here, the actual last name of the second contributor is not the final word `Zander` but the third word `Doe`
If following contributors were to be added to the list:
- "Alice Henderson"
- "Bob Johnson"
- "Michael Peterson"
- "Paul Robertson"

The insertion algorithm would place all the new contributors between the two existing contributors, as shown below:
```
<contributor name="John Doe"/>
<contributor name="Alice Henderson"/>
<contributor name="Bob Johnson"/>
<contributor name="Michael Peterson"/>
<contributor name="Paul Robertson"/>
<contributor name="Jane Ann Doe Zander"/>
..
..
```
This placement is incorrect for almost all the new names, as they end up in the wrong part of the list. Furthermore, this could create a cascade effect (if the order is not mannually corrected).

Any new contributors with similarly complex last-name structures (as Jane Ann Doe Zander) added in the future might make the insertion unreliable. 
 It might repeatedly misplace entries, ultimately defeating the purpose of maintaining a correct order.

File: conf/contributors.xml
Comment (soham30rane) [reply conf/contributors.xml:1360]: Yeah, we may mannually combine it

These two entries also existed (at different places) in original list. The sorting brought them next to each other.

File: .github/workflows/generate_changelog.yml
Comment (soham30rane) [reply .github/workflows/generate_changelog.yml:27]: Yeah works!

File: scripts/create_changelog.py
Comment (soham30rane) [reply scripts/create_changelog.py:52]: Yeah sure

File: scripts/create_changelog.py
Comment (soham30rane) [reply scripts/create_changelog.py:125]: Yes it was for debugging, will remove it

File: scripts/create_changelog.py
Comment (soham30rane) [reply scripts/create_changelog.py:175]: It was there for just testing purpose, just so that we can test on 10.5 (Since the 10.5 release is not out yet, but there are few beta's and rc's). Will remove it

File: scripts/create_changelog.py
Comment (kwankyu) [reply scripts/create_changelog.py:74]: Right. That is also awful.

Let's go with your current approach.

File: scripts/create_changelog.py
Comment (soham30rane) [reply scripts/create_changelog.py:283]: I deleted the previously generated changelog and regenerated it 🙃. Ideally, I should have reverted that particular commit instead. This is the [most recent changelog](https://github.com/sagemath/website/pull/480/commits/ce3556cad492d7bf9fcb66d52272ce545843da04#diff-a9d016a0ac7681669dc01a2aa0a29d2c86c50e9493b1ac170a508015548d30f7)

Comment (kwankyu) [reply scripts/create_changelog.py:396]: I meant "contribution". It is still "Contribution".

Comment (kwankyu) [reply scripts/create_changelog.py:396]: Sorry. I was not clear.

File: .github/workflows/generate_changelog.yml
Comment (kwankyu) [inline .github/workflows/generate_changelog.yml:39]: Would you remove the colon here?
```
"Added changelog for release ${{ inputs.release_tag }}"
```

**Comment by kwankyu** [comment]
The original formatting
```
Sage (http://www.sagemath.org) is developed by volunteers and combines
hundreds of open source packages.

The following 83 people contributed to this release. Of those, 12 made
their first contribution to Sage:
```
looks more evenly flushed at the right, compared with the slightly changed format
```
Sage (http://www.sagemath.org) is developed by volunteers and
combines hundreds of open source packages.

The following 90 people contributed to this release.
Of those, 24 made their first contribution to Sage:
```

**Comment by kwankyu** [comment]
In previous changelogs, the release manager is mentioned, like in
```
  - Yueqi Li [first contribution]

* Release manager: Volker Braun.

* We closed 549 tickets in this release.
```
but I think the period after "Braun" should be removed...

**Comment by kwankyu** [comment]
I am now happy to merge your PR. Let me know when you are done with the final touch.

**Comment by soham30rane** [comment]
I was wondering, how do we identify the release manager for a specific release? Is it mentioned explicitly somewhere, or is there a standard way to fetch this information? Or there isn't any?

File: scripts/create_changelog.py
Comment (soham30rane) [reply scripts/create_changelog.py:396]: Okay, fixed in [this commit](https://github.com/sagemath/website/pull/480/commits/519e571f7202ead3596bf44b11fea72676741084)

**Comment by kwankyu** [comment]
I hoped that the commit of a release tag contains the name of the release manager, but it is not the case.

**Comment by soham30rane** [comment]
I’m done with the final touches and ready to proceed. Thank you!

**Review by kwankyu — APPROVED** [review]
Thanks!

**Comment by soham30rane** [comment]
@kwankyu  I see the workflow failed because of branch protection (changes can only be made via pull requests). 

We might need to update the workflow to create a PR instead of directly pushing. The manual step of merging the PR seems unavoidable but aligns well with the current protection rules of this repo. What do you think?

**Comment by kwankyu** [comment]
Yes, the branch protection rule was an obstacle. It could be overcome by making the workflow to use the PAT created by me (admin). I found that Dima is also an admin since he is in the Core group. It seems that the PAT stored as a secret is secure enough.

The workflow is now working well. I manually ran the workflow for 10.1, 10.2, 10.3, 10.4 (pending).

I made some other small fixes and manual edits for 10.0 changelog. You can check them in the last commits.

**Comment by kwankyu** [comment]
There is some obscurity about the release date. The workflow gets the release date from "published at" value of the release data. However the release data is created by a workflow in sagemath/sage, which could run at worst after a few days after the release manager make a release (= push a release tag). I think there is nothing that the workflow could do about this.

**Comment by soham30rane** [comment]
The response returned by Githup API has following two dates
```
"created_at": "2024-07-19T22:37:02Z",
"published_at": "2024-07-20T00:28:33Z",
```
This example is for tag 10.4
So it seems we have to fetch the `created_at` value from the response

**Comment by kwankyu** [comment]
Thanks. Yes, it seems better. I did it in https://github.com/sagemath/website/commit/854c9c3d4cfa1c4fb3d6d8ae4108a99b1d9729bc

**Comment by dimpase** [comment]
@kwankyu - I've created a PAT, as requested. Where should it go? Can I upload/store it myself? Else, I can email it to you, encrypted. Do you have a pgp/gnu-pg/gpg key?

**Comment by kwankyu** [comment]
Dima, sorry for lately informing you about what's going on.

As explained in https://github.com/sagemath/website/pull/480#issuecomment-2508941286, I realized that since you are in the Core group,  the PATs created by you are as (perhaps more) powerful than mine. Thus and also to overcome the branch protection rule obstacle, I created SAGE_ACCESS_TOKEN, and saved it in sagemath/website as a secret.

For WEBSITE_ACCESS_TOKEN, I think it is also simpler and safer for me to create a PAT and directly save it to sagemath/sage as a secret.

**Comment by kwankyu** [comment]
Just now I ran the workflow manually for the release 10.5. 
The workflow ran well. But the generated [changelog for 10.5](https://github.com/sagemath/website/blob/master/src/changelogs/sage-10.5.txt) has a problem. 
The list of "merged PRs" for release 10.5 repeats all PRs merged in the prereleases!

What should be the solution:

(1) Fix the list of merged PRs included in the stable release (include only newly merged PRs for the release)

or

(2) Fix the workflow here to work only with the stable release (ignore prereleases)

?

See https://github.com/sagemath/sage/releases

**Comment by kwankyu** [comment]
The simplest fix seems to be

(1) Fix the list of merged PRs included in the stable release (include only newly merged PRs for the release)

For now I modified https://github.com/sagemath/sage/releases/tag/10.5 and regenerated the changelog for release 10.5 by running the workflow again. Now we have the correct [changelog](https://github.com/sagemath/website/blob/master/src/changelogs/sage-10.5.txt).

**Comment by kwankyu** [comment]
> Should I open another pr for it ?

No. Let's keep the workflow as it is. Thanks.

**Comment by kwankyu** [comment]
By the way, @haraldschilly updated the website. See the updated page: https://www.sagemath.org/development-map.html

Thank @haraldschilly and @soham30rane!

**Pull request merged**
